### PR TITLE
Add multi part aof information

### PR DESCRIPTION
**Pull request by chenyang8094**
Upstream PR link: https://github.com/redis/redis/pull/9788

### CODE REVIEW

File: topics/persistence.md [hunk 1]
```diff
@@ -34,6 +37,9 @@ AOF advantages
...
+
+  Since Redis 7.0.0, the rewrite is also completely safe. When an AOF rewrite is scheduled, Redis parent process opens a new INCR AOF file to continue writing. The child process will execute rewrite logic and generates a new BASE AOF. Redis will use a temporary manifest file to track the newly generated BASE file and INCR file. When they are ready, Redis will perform an atomic replacement operation to make this temporary manifest file take effect.At the same time, in order to avoid the problem of creating many INCR files due to the frequent failure and retries of AOF rewrite. Redis introduced an AOF rewrite limiting mechanism to ensure that AOF rewrite automatically retries at a slower and slower speed.
+
 * AOF contains a log of all the operations one after the other in an easy to understand and parse format. You can even easily export an AOF file. For instance even if you've accidentally flushed everything using the `FLUSHALL` command, as long as no rewrite of the log was performed in the meantime, you can still save your data set just by stopping the server, removing the latest command, and restarting Redis again.
 
 AOF disadvantages
```
Comment (oranagra) [inline topics/persistence.md:45]: it's odd that i don't see these disadvantages being listed:
* using a lot of memory if there are writes to the database during a rewrite (these are buffered and written to the new AOF at the end)
* all writes that arrive during rewrite are written twice.
* server could freeze writing and fsyncing these writes to the new AOF file

so maybe we should list them as disadvantages for versions before 7?

File: topics/persistence.md [hunk 2]
```diff
@@ -60,6 +66,8 @@ for faster restarts, and in the event of bugs in the AOF engine.
 
 Note: for all these reasons we'll likely end up unifying AOF and RDB into a single persistence model in the future (long term plan).
 
+Since Redis 7.0.0, when `aof-use-rdb-preamble` is set to yes, the BASE file uses RDB format.
```
Comment (oranagra) [inline topics/persistence.md:69]: how is this related to this section?
i think we can choose not to document that option at all.
or if we document it, we should mention the version where it was first introduced, and what it did before 7 too

File: topics/persistence.md [hunk 3]
```diff
@@ -116,7 +124,7 @@ Redis it will re-play the AOF to rebuild the state.
 ### Log rewriting
 
 As you can guess, the AOF gets bigger and bigger as write operations are
-performed.  For example, if you are incrementing a counter 100 times,
+performed. For example, if you are incrementing a counter 100 times,
```
Comment (oranagra) [inline topics/persistence.md:127]: these double spaces in the middle of the line are not rendered in the MD result.
they're not needed in the "source" but i'm not sure we need to bother removing them.

File: topics/persistence.md [hunk 4]
```diff
@@ -116,7 +124,7 @@ Redis it will re-play the AOF to rebuild the state.
 ### Log rewriting
```
Comment (oranagra) [inline topics/persistence.md:124]: maybe all the test you've put in the disadvantages section that describes the rewrite in 7.0 should be moved here?

Comment (oranagra) [reply topics/persistence.md:124]: that's also the place to mention preamble if we want

File: topics/persistence.md [hunk 5]
```diff
@@ -168,6 +176,8 @@ cases if you want, but the default configuration is to continue regardless
 the fact the last command in the file is not well-formed, in order to guarantee
 availability after a restart.
 
+Since Redis 7.0.0, only the last AOF file (BASE or INCR) can be truncated. Otherwise redis will exit even if aof-load-truncated is enabled.
```
Comment (oranagra) [inline topics/persistence.md:179]: i don't think we need to mention it.. logically the tail of the AOF is just the last file.

File: topics/persistence.md [hunk 6]
```diff
@@ -212,13 +222,19 @@ and a parent process.
 at the same time it writes the new changes in the old append-only file,
 so if the rewriting fails, we are safe).
 
+  Since Redis 7.0.0, the mechanism here has been changed. We have dropped in-memory buffer and open a new INCR AOF file to continue writing. If the rewriting fails finally, the old BASE and INCR files (if we have) plus this newly opened INCR file represents a full amount of data, so we are safe.
```
Comment (oranagra) [inline topics/persistence.md:225]: i think we should split the "how it works" section into two.
one (the first one) mentions redis 7, and the other mentions older versions.
they work completely different, so there's no need to add small notes about the differences.
also, it'll be easier for the reader to just read about the version he's using.

p.s. i see you indented this line with two spaces (i assume you wanted it to be part of the last bullet). but in markdown, the empty line between them will prevent that.

p.p.s, look at the README in this folder for styling guide lines on where we want line breaks to be. (i know it's confusing since the old "code" uses a different guideline)

and let's avoid the upper case terms BASE and INCR. i think they're ok in the code, but not the docs (unless we first establish these terms in some TOC, but i don't think that's needed)

File: topics/persistence.md (hunk 1 above)
Comment (chenyang8094) [reply topics/persistence.md:45]: Agree, I think it might have been deliberately circumventing this before?

File: topics/persistence.md (hunk 2 above)
Comment (chenyang8094) [reply topics/persistence.md:69]: Pay attention to the above sentence: ”for all these reasons we'll likely end up unifying AOF and RDB into a single persistence model in the future (long term plan).“ i think when `aof-use-rdb-preamble` is set to yes, we can do that, right ?

File: topics/persistence.md (hunk 3 above)
Comment (chenyang8094) [reply topics/persistence.md:127]: Yes, but I think one day someone will specifically mention a PR to remove this space?

File: topics/persistence.md (hunk 1 above)
Comment (oranagra) [reply topics/persistence.md:45]: i don't think deliberately. maybe the severity wasn't realized when this doc was first created, and didn't bother to update later.

File: topics/persistence.md (hunk 3 above)
Comment (oranagra) [reply topics/persistence.md:127]: i'll reject it if it's in my shift..  8-)
i'm usually trying to avoid changes that are unnecessary and damage the blame log.

File: topics/persistence.md (hunk 2 above)
Comment (oranagra) [reply topics/persistence.md:69]: i don't think that's it.
the section is titled `Ok, so what should I use`, and it mentions the potential for data loss, vs the overheads of AOF (constant writes, not during AOFRW)
it discusses the difference between a periodic snapshots (e.g. fork every hour), vs AOF (write every command), which has a rewrite that could be much more sparse..

it is still a long term plan for us to unify these (i.e. you can define both, and you'll end up with only AOF, but it's rewrite will be periodic).
i don't think this PR changes that.

we may choose to mention preamble somewhere (if there's a line that mentions that AOFRW is being inefficient since it generates text file), but then we'll mention the version that introduced preamble, not 7.0.
i don't think 7.0 adds a major change with regards to this section.

File: topics/persistence.md (hunk 5 above)
Comment (chenyang8094) [reply topics/persistence.md:179]: OK

File: topics/persistence.md (hunk 6 above)
Comment (chenyang8094) [reply topics/persistence.md:225]: OK

File: topics/persistence.md [hunk 7]
```diff
@@ -34,6 +37,9 @@ AOF advantages
...
 * Redis is able to automatically rewrite the AOF in background when it gets too big. The rewrite is completely safe as while Redis continues appending to the old file, a completely new one is produced with the minimal set of operations needed to create the current data set, and once this second file is ready Redis switches the two and starts appending to the new one.
+
+  Since Redis 7.0.0, the rewrite is also completely safe. When an AOF rewrite is scheduled, Redis parent process opens a new INCR AOF file to continue writing. The child process will execute rewrite logic and generates a new BASE AOF. Redis will use a temporary manifest file to track the newly generated BASE file and INCR file. When they are ready, Redis will perform an atomic replacement operation to make this temporary manifest file take effect.At the same time, in order to avoid the problem of creating many INCR files due to the frequent failure and retries of AOF rewrite. Redis introduced an AOF rewrite limiting mechanism to ensure that AOF rewrite automatically retries at a slower and slower speed.
```
Comment (yoav-steinberg) [inline topics/persistence.md:41]:
```suggestion
  Since Redis 7.0.0, the rewrite is also completely safe. When an AOF rewrite is scheduled, The Redis parent process opens a new INCR AOF file to continue writing. The child process will execute rewrite logic and generates a new BASE AOF. Redis will use a temporary manifest file to track the newly generated BASE file and INCR file. When they are ready, Redis will perform an atomic replacement operation to make this temporary manifest file take effect. At the same time, in order to avoid the problem of creating many INCR files in case of repeated failures and retries of an AOF rewrite, Redis introduces an AOF rewrite limiting mechanism to ensure that AOF rewrite automatically retries at a slower and slower speed.
```

File: topics/persistence.md [hunk 8]
```diff
@@ -34,6 +37,9 @@ AOF advantages
...
+
```
[the remark this answers is not in the document]
Comment (yoav-steinberg) [reply topics/persistence.md:42]: Should probably move to the `log rewriting` section below.

File: topics/persistence.md [hunk 9]
```diff
@@ -6,7 +6,10 @@ Redis Persistence
...
-* **AOF** (Append Only File):  The AOF persistence logs every write operation received by the server, that will be played again at server startup, reconstructing the original dataset. Commands are logged using the same format as the Redis protocol itself, in an append-only fashion. Redis is able to rewrite the log in the background when it gets too big.
+* **AOF** (Append Only File): The AOF persistence logs every write operation received by the server, that will be played again at server startup, reconstructing the original dataset. Commands are logged using the same format as the Redis protocol itself, in an append-only fashion. Redis is able to rewrite the log in the background when it gets too big.    
+
+  Since Redis 7.0.0, Redis supports the Multi Part AOF mechanism. That is, the original single AOF file is split into BASE file (at most one) and INCR file (there may be more than one). The BASE file represents an initial (RDB or AOF format) snapshot of the data present when the AOF is rewritten. The INCR file contains incremental commands since the last base AOF file was created. All these files are tracked by a manifest file.
```
[the remark this answers is not in the document]
Comment (chenyang8094) [reply topics/persistence.md:11]: OK

File: topics/persistence.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (chenyang8094) [reply topics/persistence.md:42]: ok

File: topics/persistence.md (hunk 2 above)
Comment (chenyang8094) [reply topics/persistence.md:69]: removed

File: topics/persistence.md (hunk 3 above)
Comment (chenyang8094) [reply topics/persistence.md:127]: OK

File: topics/persistence.md (hunk 4 above)
Comment (chenyang8094) [reply topics/persistence.md:124]: OK

File: topics/persistence.md (hunk 7 above)
Comment (chenyang8094) [reply topics/persistence.md:41]: OK

[the remark this answers is not in the document]
Comment (chenyang8094) [reply topics/persistence.md]: OK

File: topics/persistence.md [hunk 10]
```diff
@@ -33,14 +39,16 @@ AOF advantages
...
 * AOF contains a log of all the operations one after the other in an easy to understand and parse format. You can even easily export an AOF file. For instance even if you've accidentally flushed everything using the `FLUSHALL` command, as long as no rewrite of the log was performed in the meantime, you can still save your data set just by stopping the server, removing the latest command, and restarting Redis again.
 
 AOF disadvantages
 ---
 
 * AOF files are usually bigger than the equivalent RDB files for the same dataset.
 * AOF can be slower than RDB depending on the exact fsync policy. In general with fsync set to *every second* performance is still very high, and with fsync disabled it should be exactly as fast as RDB even under high load. Still RDB is able to provide more guarantees about the maximum latency even in the case of an huge write load.
+* Using a lot of memory if there are writes to the database during a rewrite (these are buffered in memory and written to the new AOF at the end).
```
Comment (yoav-steinberg) [inline topics/persistence.md:49]: Isn't this removed with the new muti-part AOF mechanism? Now we write the new writes to the new incremental file, don't we?

File: topics/persistence.md [hunk 11]
```diff
@@ -33,14 +39,16 @@ AOF advantages
...
+* All writes that arrive during rewrite are written twice.
```
Comment (yoav-steinberg) [inline topics/persistence.md:50]: In normal operation this shouldn't happen, they are written only to the new incremental file aren't they?

File: topics/persistence.md [hunk 12]
```diff
@@ -33,14 +39,16 @@ AOF advantages
...
+* Redis could freeze writing and fsyncing these writes to the new AOF file at the end.
```
Comment (yoav-steinberg) [inline topics/persistence.md:51]: What do you mean? If we continue writing as normal to the new incr file during the rewrite then why would we freeze?

File: topics/persistence.md [hunk 13]
```diff
@@ -121,6 +129,14 @@ you'll end up with a single key in your dataset containing the final
 value, but 100 entries in your AOF. 99 of those entries are not needed
 to rebuild the current state.
 
+The rewrite is completely safe as while Redis continues appending to the old file, a completely new one is produced with the minimal set of operations needed to create the current data set, and once this second file is ready Redis switches the two and starts appending to the new one.
```
Comment (yoav-steinberg) [inline topics/persistence.md:132]:
```suggestion
The rewrite is completely safe. While Redis continues appending to the old file, a completely new one is produced with the minimal set of operations needed to create the current data set, and once this second file is ready Redis switches the two and starts appending to the new one.
```

File: topics/persistence.md [hunk 14]
```diff
@@ -121,6 +129,14 @@ you'll end up with a single key in your dataset containing the final
...
+
+Since Redis 7.0.0, when an AOF rewrite is scheduled, The Redis parent process opens a new incremental AOF file to continue writing. The child process will execute rewrite logic and generates a new base AOF.
+Redis will use a temporary manifest file to track the newly generated base file and incremental file.
+When they are ready, Redis will perform an atomic replacement operation to make this temporary manifest file take effect.
+At the same time, in order to avoid the problem of creating many incr files in case of repeated failures and retries of an AOF rewrite.
```
Comment (yoav-steinberg) [inline topics/persistence.md:137]:
```suggestion
In order to avoid the problem of creating many incr files in case of repeated failures and retries of an AOF rewrite,
```

File: topics/persistence.md [hunk 15]
```diff
@@ -121,6 +129,14 @@ you'll end up with a single key in your dataset containing the final
...
+Redis introduces an AOF rewrite limiting mechanism to ensure that AOF rewrite automatically retries at a slower and slower speed.
```
Comment (yoav-steinberg) [inline topics/persistence.md:138]:
```suggestion
Redis introduces an AOF rewrite limiting mechanism to ensure that AOF rewrite automatically retries at a slower and slower rate.
```

File: topics/persistence.md [hunk 16]
```diff
@@ -203,6 +219,22 @@ corruption happened to be in the initial part of the file.
...
 snapshotting.  This is how it works:
 
+**Redis >= 7.0**
+
+* Redis [forks](http://linux.die.net/man/2/fork), so now we have a child
+and a parent process.
+
+* The child starts writing the new AOF in a temporary file.
+
+* The parent open a new incr AOF file to continue writing. If the rewriting fails finally, the old base and incr files (if we have) plus this newly opened incr file represents a full amount of data, so we are safe.
```
Comment (yoav-steinberg) [inline topics/persistence.md:229]:
```suggestion
* The parent opens a new increments AOF file to continue writing updates. If the rewriting fails, the old base and increments files (if we have) plus this newly opened increments file represents the complete updated dataset, so we are safe.
```

File: topics/persistence.md [hunk 17]
```diff
@@ -203,6 +219,22 @@ corruption happened to be in the initial part of the file.
...
+  
+* When the child is done rewriting the file, the parent gets a signal,
+and use a temporary manifest to track the newly opened incr file and sub-process generated base files.
```
Comment (yoav-steinberg) [inline topics/persistence.md:232]:
```suggestion
and use a temporary manifest file to track the newly opened increments file and child generated base file.
```

File: topics/persistence.md [hunk 18]
```diff
@@ -203,6 +219,22 @@ corruption happened to be in the initial part of the file.
...
+
+* Profit! Now Redis do an atomic exchange to the manifest pointer in the memory, so that the result of this AOF rewrite is visible to Redis.
```
Comment (yoav-steinberg) [inline topics/persistence.md:234]: Isn't the atomic operation renaming the manifest file, not merely a change of a pointer in memory?

File: topics/persistence.md [hunk 19]
```diff
@@ -291,7 +323,11 @@ running. This is what we suggest:
 If you run a Redis instance with only AOF persistence enabled, you can still
 copy the AOF in order to create backups. The file may lack the final part
 but Redis will be still able to load it (see the previous sections about
-truncated AOF files).
+truncated AOF files).  
+
+Since Redis 7.0.0, all the base, incr and manifest files will be placed in a file directory determined by the `appendddirname` configuration.
+so the best suggestion is to copy the entire directory directly when backup the AOF, instead of copying some of the files,
```
Comment (yoav-steinberg) [inline topics/persistence.md:329]:
```suggestion
so the best suggestion is to copy the entire directory when backing up AOF persistence.
```

File: topics/persistence.md [hunk 20]
```diff
@@ -291,7 +323,11 @@ running. This is what we suggest:
 If you run a Redis instance with only AOF persistence enabled, you can still
 copy the AOF in order to create backups. The file may lack the final part
 but Redis will be still able to load it (see the previous sections about
-truncated AOF files).
+truncated AOF files).  
+
+Since Redis 7.0.0, all the base, incr and manifest files will be placed in a file directory determined by the `appendddirname` configuration.
+so the best suggestion is to copy the entire directory directly when backup the AOF, instead of copying some of the files, 
+because Redis will do strictly check when starts, once some files are lost or messy, Redis will refuse to start.
```
Comment (yoav-steinberg) [inline topics/persistence.md:330]:
```suggestion
```

File: topics/persistence.md (hunk 10 above)
Comment (oranagra) [reply topics/persistence.md:49]: i asked to add these 3 disadvantages, but it should be clear that these are issues relevant only to versions before 7.
maybe make a sub-section for pre-7.0 disadvantages?

File: topics/persistence.md (hunk 13 above)
Comment (oranagra) [reply topics/persistence.md:132]: the line is quite long. it'l be easier for review, and future updates if we split it after some commas or alike, so adding a line will not create such a big diff.

File: topics/persistence.md (hunk 10 above)
Comment (chenyang8094) [reply topics/persistence.md:49]: Yes, i think i should add **Redis < 7.0** and  **Redis >= 7.0** to splite them?

Comment (oranagra) [reply topics/persistence.md:49]: not sure if we should split it all..
there could be some generic disadvantages that apply to all.
and then some version specific disadvantages

File: topics/persistence.md (hunk 13 above)
Comment (chenyang8094) [reply topics/persistence.md:132]: This is the previous document, do we need to divide them into shorter lines now?

File: topics/persistence.md (hunk 14 above)
Comment (chenyang8094) [reply topics/persistence.md:137]: OK

File: topics/persistence.md (hunk 15 above)
Comment (chenyang8094) [reply topics/persistence.md:138]: OK

File: topics/persistence.md (hunk 16 above)
Comment (chenyang8094) [reply topics/persistence.md:229]: OK

File: topics/persistence.md (hunk 17 above)
Comment (chenyang8094) [reply topics/persistence.md:232]: OK

File: topics/persistence.md (hunk 19 above)
Comment (chenyang8094) [reply topics/persistence.md:329]: OK

File: topics/persistence.md (hunk 20 above)
Comment (chenyang8094) [reply topics/persistence.md:330]: OK

File: topics/persistence.md (hunk 10 above)
Comment (chenyang8094) [reply topics/persistence.md:49]: @oranagra I think we should delete this sentence: `In the past we experienced rare bugs in specific commands (for instance there was one involving blocking commands like BRPOPLPUSH) causing the AOF produced to not reproduce exactly the same dataset on reloading. These bugs are rare and we have tests in the test suite creating ran`，We should talk about the disadvantages of AOF, not its previous bugs, WDYT?

File: topics/persistence.md (hunk 18 above)
Comment (chenyang8094) [reply topics/persistence.md:234]: I updated the above sentence

File: topics/persistence.md (hunk 13 above)
Comment (oranagra) [reply topics/persistence.md:132]: sorry, it looked new since it was moved.
i guess that since we moved it, we can take the opportunity and break it to shorter lines.
it's not a hard rule.. just a guideline for new parts we add.

Comment (chenyang8094) [reply topics/persistence.md:132]: OK

File: topics/persistence.md (hunk 10 above)
Comment (oranagra) [reply topics/persistence.md:49]: yes, i agree it's outdated...
i also think we can afford to delete this odd thing:
```
We have never had a single report from users about an AOF corruption that was detected in the real world.
  ```

Comment (chenyang8094) [reply topics/persistence.md:49]: agree

File: topics/persistence.md [hunk 21]
```diff
@@ -33,18 +39,19 @@ AOF advantages
...
-* In the past we experienced rare bugs in specific commands (for instance there was one involving blocking commands like `BRPOPLPUSH`) causing the AOF produced to not reproduce exactly the same dataset on reloading. These bugs are rare and we have tests in the test suite creating random complex datasets automatically and reloading them to check everything is fine. However, these kind of bugs are almost impossible with RDB persistence. To make this point more clear: the Redis AOF works by incrementally updating an existing state, like MySQL or MongoDB does, while the RDB snapshotting creates everything from scratch again and again, that is conceptually more robust. However -
-  1) It should be noted that every time the AOF is rewritten by Redis it is recreated from scratch starting from the actual data contained in the data set, making resistance to bugs stronger compared to an always appending AOF file (or one rewritten reading the old AOF instead of reading the data in memory).
-  2) We have never had a single report from users about an AOF corruption that was detected in the real world.
 
+**Redis < 7.0**
```
Comment (yoav-steinberg) [inline topics/persistence.md:50]: Missing newline (breaks markdown bullet formatting)
```suggestion
**Redis < 7.0**

```

File: topics/persistence.md [hunk 22]
```diff
@@ -33,18 +39,19 @@ AOF advantages
...
+* Using a lot of memory if there are writes to the database during a rewrite (these are buffered in memory and written to the new AOF at the end).
```
Comment (yoav-steinberg) [inline topics/persistence.md:51]:
```suggestion
* AOF can use a lot of memory if there are writes to the database during a rewrite (these are buffered in memory and written to the new AOF at the end).
```

File: topics/persistence.md [hunk 23]
```diff
@@ -33,18 +39,19 @@ AOF advantages
...
+* All writes that arrive during rewrite are written twice.
+* Redis could freeze writing and fsyncing these writes to the new AOF file at the end.
```
Comment (yoav-steinberg) [inline topics/persistence.md:53]:
```suggestion
* Redis could freeze writing and fsyncing these write commands to the new AOF file at the end.
```

File: topics/persistence.md [hunk 24]
```diff
@@ -33,18 +39,19 @@ AOF advantages
...
-  1) It should be noted that every time the AOF is rewritten by Redis it is recreated from scratch starting from the actual data contained in the data set, making resistance to bugs stronger compared to an always appending AOF file (or one rewritten reading the old AOF instead of reading the data in memory).
-  2) We have never had a single report from users about an AOF corruption that was detected in the real world.
 
+**Redis < 7.0**
+* Using a lot of memory if there are writes to the database during a rewrite (these are buffered in memory and written to the new AOF at the end).
+* All writes that arrive during rewrite are written twice.
```
Comment (yoav-steinberg) [inline topics/persistence.md:52]:
```suggestion
* All write commands that arrive during rewrite are written twice.
```

File: topics/persistence.md [hunk 25]
```diff
@@ -6,7 +6,13 @@ Redis Persistence
...
+* **AOF** (Append Only File): The AOF persistence logs every write operation received by the server, that will be played again at server startup, reconstructing the original dataset. Commands are logged using the same format as the Redis protocol itself, in an append-only fashion. Redis is able to rewrite the log in the background when it gets too big.    
+
+  Since Redis 7.0.0, Redis supports the Multi Part AOF mechanism.
+  That is, the original single AOF file is split into base file (at most one) and incremental files (there may be more than one).
+  The base file represents an initial (RDB or AOF format) snapshot of the data present when the AOF is rewritten.
+  The incremental files contains incremental commands since the last base AOF file was created. All these files are put in a separate folder and are tracked by a manifest file.
```
Comment (yoav-steinberg) [inline topics/persistence.md:14]:
```suggestion
  <br/>Since Redis 7.0.0, Redis supports a multi part AOF mechanism.
  That is, the original single AOF file is split into base file (at most one) and incremental files (there may be more than one).
  The base file represents an initial (RDB or AOF format) snapshot of the data present when the AOF is rewritten.
  The incremental files contains incremental commands since the last base AOF file was created. All these files are put in a separate directory and are tracked by a manifest file.
```

File: topics/persistence.md [hunk 26]
```diff
@@ -121,6 +128,18 @@ you'll end up with a single key in your dataset containing the final
...
+The rewrite is completely safe.
+While Redis continues appending to the old file,
+a completely new one is produced with the minimal set of operations needed to create the current data set,
+and once this second file is ready Redis switches the two and starts appending to the new one.
+
+Since Redis 7.0.0, when an AOF rewrite is scheduled, The Redis parent process opens a new incremental AOF file to continue writing.
+The child process will execute rewrite logic and generates a new base AOF.
+Redis will use a temporary manifest file to track the newly generated base file and incremental file.
+When they are ready, Redis will perform an atomic replacement operation to make this temporary manifest file take effect.
+In order to avoid the problem of creating many incr files in case of repeated failures and retries of an AOF rewrite.
```
Comment (yoav-steinberg) [inline topics/persistence.md:140]:
```suggestion
In order to avoid the problem of creating many incremental files in case of repeated failures and retries of an AOF rewrite.
```

File: topics/persistence.md [hunk 27]
```diff
@@ -203,6 +222,25 @@ corruption happened to be in the initial part of the file.
...
 
+**Redis >= 7.0**
+
+* Redis [forks](http://linux.die.net/man/2/fork), so now we have a child
+and a parent process.
+
+* The child starts writing the new AOF in a temporary file.
+
+* The parent opens a new increments AOF file to continue writing updates.
+  If the rewriting fails, the old base and increments files (if we have) plus this newly opened increments file represents the complete updated dataset,
```
Comment (yoav-steinberg) [inline topics/persistence.md:233]:
```suggestion
  If the rewriting fails, the old base and increments files (if we have) plus this newly opened increments file represent the complete updated dataset,
```

File: topics/persistence.md [hunk 28]
```diff
@@ -203,6 +222,25 @@ corruption happened to be in the initial part of the file.
...
+  so we are safe.
+  
+* When the child is done rewriting the file, the parent gets a signal,
```
Comment (yoav-steinberg) [inline topics/persistence.md:236]:
```suggestion
* When the child is done rewriting the base file, the parent gets a signal,
```

File: topics/persistence.md [hunk 29]
```diff
@@ -203,6 +222,25 @@ corruption happened to be in the initial part of the file.
...
+and use newly opened increment file and child generated base file to build an temp manifest,
```
Comment (yoav-steinberg) [inline topics/persistence.md:237]:
```suggestion
and uses the newly opened increment file and child generated base file to build a temp manifest,
```

File: topics/persistence.md [hunk 30]
```diff
@@ -203,6 +222,25 @@ corruption happened to be in the initial part of the file.
...
+and persist it.
+
+* Profit! Now Redis do an atomic exchange to the manifest pointer in the memory, so that the result of this AOF rewrite is visible to Redis.
```
Comment (yoav-steinberg) [inline topics/persistence.md:240]:
```suggestion
* Profit! Now Redis does an atomic exchange of the manifest files so that the result of this AOF rewrite takes effect. Redis also cleans up the old base file, manifest file and any unused increment files.
```

File: topics/persistence.md [hunk 31]
```diff
@@ -291,7 +329,10 @@ running. This is what we suggest:
 If you run a Redis instance with only AOF persistence enabled, you can still
 copy the AOF in order to create backups. The file may lack the final part
 but Redis will be still able to load it (see the previous sections about
-truncated AOF files).
+truncated AOF files).  
+
+Since Redis 7.0.0, all the base, incr and manifest files will be placed in a file directory determined by the `appendddirname` configuration.
```
Comment (yoav-steinberg) [inline topics/persistence.md:334]:
```suggestion
Since Redis 7.0.0, all the base, increment and manifest files will be placed in a file directory determined by the `appendddirname` configuration.
```

File: topics/persistence.md (hunk 21 above)
Comment (chenyang8094) [reply topics/persistence.md:50]: OK

File: topics/persistence.md (hunk 22 above)
Comment (chenyang8094) [reply topics/persistence.md:51]: OK

File: topics/persistence.md (hunk 24 above)
Comment (chenyang8094) [reply topics/persistence.md:52]: OK

File: topics/persistence.md (hunk 23 above)
Comment (chenyang8094) [reply topics/persistence.md:53]: OK

File: topics/persistence.md (hunk 25 above)
Comment (chenyang8094) [reply topics/persistence.md:14]: OK

File: topics/persistence.md (hunk 27 above)
Comment (chenyang8094) [reply topics/persistence.md:233]: OK

File: topics/persistence.md [hunk 32]
```diff
@@ -203,6 +222,25 @@ corruption happened to be in the initial part of the file.
 Log rewriting uses the same copy-on-write trick already in use for
 snapshotting.  This is how it works:
 
+**Redis >= 7.0**
+
+* Redis [forks](http://linux.die.net/man/2/fork), so now we have a child
+and a parent process.
+
+* The child starts writing the new AOF in a temporary file.
```
Comment (oranagra) [inline topics/persistence.md:230]:
```suggestion
* The child starts writing the new base AOF in a temporary file.
```

File: topics/persistence.md [hunk 33]
```diff
@@ -203,6 +222,25 @@ corruption happened to be in the initial part of the file.
...
+
+* The parent opens a new increments AOF file to continue writing updates.
+  If the rewriting fails, the old base and increments files (if we have) plus this newly opened increments file represent the complete updated dataset,
+  so we are safe.
```
Comment (oranagra) [inline topics/persistence.md:234]: "increments file" sounds odd, i'd refer "incremental file"
but i see you also have "increments files", could this be "incremental files"?
```suggestion
* The parent opens a new increments AOF file to continue writing updates.
  If the rewriting fails, the old base and increments files (if there are any) plus this newly opened increments file represent the complete updated dataset,
  so we are safe.
```

File: topics/persistence.md [hunk 34]
```diff
@@ -203,6 +222,25 @@ corruption happened to be in the initial part of the file.
...
+  
+* When the child is done rewriting the base file, the parent gets a signal,
+and uses the newly opened increment file and child generated base file to build an temp manifest,
+and persist it.
```
Comment (oranagra) [inline topics/persistence.md:238]: maybe it would be clearer if we point out that there are no usually a base file and two incremental files (+manifest).

**Review by oranagra — APPROVED** [review]

File: topics/persistence.md (hunk 33 above)
Comment (yoav-steinberg) [reply topics/persistence.md:234]: > "increments file" sounds odd, i'd refer "incremental file"

I think the best is "increment files", i'll fix

File: topics/persistence.md [hunk 35]
```diff
@@ -203,6 +222,25 @@ corruption happened to be in the initial part of the file.
...
+and a parent process.
+
+* The child starts writing the new base AOF in a temporary file.
+
+* The parent opens a new increments AOF file to continue writing updates.
+  If the rewriting fails, the old base and increments files (if there are any) plus this newly opened increments file represent the complete updated dataset,
+  so we are safe.
+  
+* When the child is done rewriting the base file, the parent gets a signal,
+and uses the newly opened increment file and child generated base file to build an temp manifest,
```
Comment (yoav-steinberg) [inline topics/persistence.md:237]:
```suggestion
and uses the newly opened increment file and child generated base file to build a temp manifest,
```

File: topics/persistence.md [hunk 36]
```diff
@@ -203,6 +222,25 @@ corruption happened to be in the initial part of the file.
...
 
+**Redis >= 7.0**
+
+* Redis [forks](http://linux.die.net/man/2/fork), so now we have a child
+and a parent process.
+
+* The child starts writing the new base AOF in a temporary file.
+
+* The parent opens a new increments AOF file to continue writing updates.
+  If the rewriting fails, the old base and increments files (if there are any) plus this newly opened increments file represent the complete updated dataset,
```
Comment (yoav-steinberg) [inline topics/persistence.md:233]:
```suggestion
  If the rewriting fails, the old base and increment files (if there are any) plus this newly opened increments file represent the complete updated dataset,
```

File: topics/persistence.md (hunk 34 above)
Comment (yoav-steinberg) [reply topics/persistence.md]: At this stage there's normally one new inc file, one old inc file, one new base file and one old base file. When done writing the new base file we just use it and the **new** inc file. This is what it says here. What do you think is missing?

Comment (chenyang8094) [reply topics/persistence.md]: before AOFRW, we may have: `BASE1 + INCR1 + MANIFEST`
start AOFRW, we have:  `BASE1 + INCR1 + INCR2 + MANIFEST`
child process exit normally AOFRW (in doneHandler), we have:  `BASE1 + INCR1 + INCR2 + MANIFEST +  BASE2 `
AOFRW finish (del history), we have: `BASE2 + INCR2`
